### PR TITLE
Support Django 1.8(.3)

### PR DESCRIPTION
**Pull request by awwad**
Currently, SeattleTestbed/clearinghouse:master runs on Django version 1.6, and does not support Django 1.7 or later. The changes here allow us to run on Django versions through at least Django 1.8.3 (likely through 1.8.x).

See [Django deprecation/removal timeline](https://docs.djangoproject.com/en/1.9/internals/deprecation/) for some of the changes that affected our codebase. In particular, I dealt with:
  -   legacy db transaction control removal
  -   removals of content_type argument support
  -   changes to app loading procedure
  -   switch to migration-based db model
  -   (fill in rest)

Some changes were made by user Kaushik1091, and some by me.

Testing:
[The Clearinghouse installation procedure is here](https://seattle.poly.edu/wiki/ClearinghouseInstallation). While it is updated, Django 1.7+-exclusive instructions are not noted there (In particular, instead of manage.py syncdb, you should employ make manage.py makemigrations followed by manage.py migrate).
Existing code in the pull request has been tested by me on an Ubuntu VM running Django 1.8.3 (also a bit on 1.7 and 1.6.3, though less thoroughly). The clearinghouse website appears to function fully. I've created users, downloaded keys, downloaded donator installers and installed them, seen the donations be registered and carved up by the clearinghouse, requested vessels, and received them. More testing is merited, but it seems to be functional to my eyes.

At present, some of the comments are a bit sloppy or lacking.

### CODE REVIEW

**Comment by awwad** [comment]
Fixes #152

**Comment by awwad** [comment]
Cleaned up :P

File: website/html/templates/common/installers.html [hunk 1]
```diff
@@ -61,7 +61,7 @@
 						<a href="{% url 'linux_installer' username %}">Download installer for Linux</a>
 						<p>
 							<strong>Instructions:</strong>
-							<br />(Note: You must have Python 2.5 installed)
+							<br />(Note: You must have Python 2.5, 2.6, or 2.7 installed)
```
Comment (lukpueh) [inline website/html/templates/common/installers.html]: 👍

File: website/html/templates/common/installers.html [hunk 2]
```diff
@@ -28,7 +28,7 @@
 						<a href="{% url 'mac_installer' username %}">Download installer for Mac</a>
 						<p>
 							<strong>Instructions:</strong>
-							<br />(Note: You must have Python 2.5 installed)
+							<br />(Note: You must have Python 2.5, 2.6, or 2.7 installed)
```
Comment (lukpueh) [inline website/html/templates/common/installers.html]: 👍

File: website/html/templates/common/geni_base.html [hunk 3]
```diff
@@ -30,29 +29,33 @@
...
+                <!--<span id="caption">Learn about <a href="{{ TESTBED_URL }}">{{ TESTBED }}</a></span> -->
+            </div>
+        {% endblock header%}
+
+        {% block navigation %}{% endblock navigation %}
+        {% block content %}{% endblock content %}
+
+    </body>
+</html> 
+
```
Comment (lukpueh) [inline website/html/templates/common/geni_base.html]: Looks like some arbitrary reformatting. Not really a candidate for this PR.

Btw. I think server side comments `{% comment %}{% endcomment %}` should be preferred over client-side comments `<!-- -->`

File: website/html/views.py [hunk 4]
```diff
@@ -868,7 +873,9 @@ def priv_key(request):
   except LoggedInButFailedGetGeniUserError:
     return _show_failed_get_geniuser_page(request)
   
-  response = HttpResponse(user.user_privkey, mimetype='text/plain')
+  # Fixed for Issue #152, compatibility with Django 1.8+ requires shift from
+  # mime_type argument to content_type argument.
```
Comment (lukpueh) [inline website/html/views.py]: [True](https://docs.djangoproject.com/en/dev/internals/deprecation/#deprecation-removed-in-1-7)

File: website/html/views.py [hunk 5]
```diff
@@ -868,7 +873,9 @@ def priv_key(request):
   except LoggedInButFailedGetGeniUserError:
     return _show_failed_get_geniuser_page(request)
   
-  response = HttpResponse(user.user_privkey, mimetype='text/plain')
+  # Fixed for Issue #152, compatibility with Django 1.8+ requires shift from
+  # mime_type argument to content_type argument.
+  response = HttpResponse(user.user_privkey, content_type='text/plain')
```
Comment (lukpueh) [inline website/html/views.py]: [True](https://docs.djangoproject.com/en/dev/internals/deprecation/#deprecation-removed-in-1-7) too

File: website/html/views.py [hunk 6]
```diff
@@ -885,7 +892,9 @@ def pub_key(request):
   except LoggedInButFailedGetGeniUserError:
     return _show_failed_get_geniuser_page(request)
   
-  response = HttpResponse(user.user_pubkey, mimetype='text/plain')
+  # Fixed for Issue #152, compatibility with Django 1.8+ requires shift from
+  # mime_type argument to content_type argument.
```
Comment (lukpueh) [inline website/html/views.py]: [True](https://docs.djangoproject.com/en/dev/internals/deprecation/#deprecation-removed-in-1-7) too too

File: website/settings.py [hunk 7]
```diff
@@ -196,9 +195,10 @@
   'django.contrib.messages',
   # Needed for OpenID/OAuth login so must be listed. 
   'social_auth',
+  'clearinghouse.website.control',
 
   # We have our maindb model defined here, so it must be listed.
-  'clearinghouse.website.control',
+  # 'clearinghouse.website.control.models',
```
Comment (lukpueh) [inline website/settings.py]: If it must be listed here, why is it commented out?

File: website/settings.py [hunk 8]
```diff
@@ -215,7 +215,7 @@
   #'social_auth.backends.browserid.BrowserIDBackend',
   #'social_auth.backends.contrib.live.LiveBackend',
   # Django default this is always needed and must always be last.
-  'django.contrib.auth.backends.ModelBackend',	    
+  'django.contrib.auth.backends.ModelBackend',
```
Comment (lukpueh) [inline website/settings.py]: Two trailing spaces instead of one?

File: website/control/models.py [hunk 9]
```diff
@@ -31,36 +32,37 @@
...
 
 # If this is a modern-enough version of django to support specifying a function
 # to be called on database connection creation, then have it call init_maindb()
 # at that time. This is to help prevent init_maindb() from accidentally not
 # being called when it should be.
-if django.VERSION >= (1,1):
-  # connection_created only exists with django >= 1.1
-  import django.db.backends.signals
-  django.db.backends.signals.connection_created.connect(_prepare_newly_created_db_connection)
+if django.VERSION >= (1.1):
```
Comment (lukpueh) [inline website/control/models.py]: @aaaaalbert pointed out that we can safely get rid of this version check. If someone uses django <= 1.1 we really can't help. Apart from this line I didn't spot any other exaggerated compatibility legacy.

File: backend/backend_daemon.py [hunk 10]
```diff
@@ -337,8 +337,25 @@ def cleanup_vessels():
...
+  # 1. Remove enter_transaction_management() line (Functionality removed in
+  #    Django 1.8. Expecting autocommit to work in its absence.)
+  # 2. Add code to handle calling django.setup() in the appropriate conditions.
+
+  ## Start a transaction management.
+  #django.db.transaction.enter_transaction_management()
+
+  # If we are in a django version that has django.setup() (1.7+?), run django.setup()
+  # This is needed for django compatibility.
+  # So that it works, undo the repy custom patching of the type builtin so that django can use it. We'll turn it back on after.
```
Comment (aaaaalbert) [inline backend/backend_daemon.py]: Also, let's not worry about Django versions that don't have `django.setup`. Backwards compatibility is of no use here. Once we require Django 1.8 to be installed for a CH, we cease supporting previous versions. (Side note: The point is more difficult to make for the Seattle "client" software where we have a large installed base with an old version of Python.)

File: backend/backend_daemon.py [hunk 11]
```diff
@@ -491,6 +508,19 @@ def sync_user_keys_of_vessels():
...
+  # undo the repy custom patching of the type builtin so that django can use it.
+  # We'll turn it back on after.
+  if hasattr(django, 'setup'):
+    temp_type = __builtins__.type
+    import safe
+    __builtins__.type = safe._type
+    django.setup()
+    __builtins__.type = temp_type
+  # end of part of fix for Issue #152
+
```
Comment (aaaaalbert) [inline backend/backend_daemon.py]: See above --- we require a suitable Django version with `django.setup` from now on.

File: node_state_transitions/node_transition_lib.py [hunk 12]
```diff
@@ -65,6 +65,18 @@
...
 from clearinghouse.website import settings
 
+# Beginning of part of fix for Issue #152
+# Perform django setup (required in Django 1.7+ (possibly 1.6.?+)) to make
+# db/models available. We need to do this if we are using a django version
+# that has django.setup() (1.7+?) This should happen before the repy patching
+# of builtins, which likely occurs in the add_dy_support call in the next line
+# of code. django.setup() expects to be able to use the builtin keyword
+# 'type', which is patched by repy, so we should do this here, before that
+# patching.
```
Comment (aaaaalbert) [inline node_state_transitions/node_transition_lib.py]: See above --- don't provide backwards compatibility.

File: polling/check_active_db_nodes.py [hunk 13]
```diff
@@ -65,6 +65,19 @@ def main():
...
   lockserver_handle = lockserver.create_lockserver_handle()
 
+  # Beginning of part of fix for Issue #152
+  # Perform django setup (required in Django 1.7+ (possibly 1.6.?+)) to make
+  # db/models available. We need to do this if we are using a django version
+  # that has django.setup() (1.7+?) This should happen before the repy patching
+  # of builtins, which likely occurs in the add_dy_support call in the next line
+  # of code. django.setup() expects to be able to use the builtin keyword
+  # 'type', which is patched by repy, so we should do this here, before that
+  # patching.
```
Comment (aaaaalbert) [inline polling/check_active_db_nodes.py]: See above --- don't provide backwards compatibility.

**Comment by aaaaalbert** [comment]
I'm working to address @lukpueh's comments now.

File: common/api/maindb.py [hunk 14]
```diff
@@ -219,21 +219,22 @@ def create_user(username, password, email, affiliation, user_pubkey, user_privke
...
+                              user_privkey=user_privkey, donor_pubkey=donor_pubkey,
+                              usable_vessel_port=port,
+                              free_vessel_credits=DEFAULT_FREE_VESSEL_CREDITS)
+          # Set the password using this function so that it gets hashed by django.
+          geniuser.set_password(password)
+          geniuser.save()
+
+          regenerate_api_key(geniuser)
     
   except:
```
Comment (aaaaalbert) [inline common/api/maindb.py]: (Well-spotted! I've added issue #175 for this.)

File: polling/check_active_db_nodes.py (hunk 13 above)
Comment (aaaaalbert) [inline polling/check_active_db_nodes.py]: Hmmm, this script doesn't seem to import `repyportability`/`safe`/`repy`directly  at all, and it doesn't attempt to restore an overridden `type` either. Thus, there's no point in keeping that comment block.

File: node_state_transitions/node_transition_lib.py (hunk 12 above)
Comment (aaaaalbert) [inline node_state_transitions/node_transition_lib.py]: In contrast to `polling/check_active_db_nodes.py`, this here `node_state_transitions/node_transition_lib.py` *does* import `repyportability` which indirectly overrides `type`. However, *again* no attempts are made to undo the override. (I thus wonder if it's really necessary to restore Python's `type` before running `django.setup()`).

File: website/control/models.py [hunk 15]
```diff
@@ -426,7 +428,7 @@ class ActionLogVesselDetails(models.Model):
 
   # To know the port despite the fact that this may change for the
   # node record itself.
-  node_port = models.IntegerField("Node port", max_length=100, db_index=True)
+  node_port = models.IntegerField("Node port", db_index=True)
```
Comment (aaaaalbert) [inline website/control/models.py]: Hmmmmm, this change goes back toe9ef43b0bb44b879fdd7565d4da93d81789a2e50 and isn't motivated clearly. I'm leaning towards not accepting it.

File: backend/backend_daemon.py [hunk 16]
```diff
@@ -337,8 +337,25 @@ def cleanup_vessels():
   
   log.info("[cleanup_vessels] cleanup thread started.")
 
-  # Start a transaction management.
-  django.db.transaction.enter_transaction_management()
+  # Beginning of part of fix for Issue #152
+  # 1. Remove enter_transaction_management() line (Functionality removed in
+  #    Django 1.8. Expecting autocommit to work in its absence.)
```
Comment (aaaaalbert) [inline backend/backend_daemon.py]: Fixed in c849e0723cecd9efc3843c4a1f3c3132f5b29a60.

File: common/api/maindb.py [hunk 17]
```diff
@@ -159,7 +159,7 @@ def init_maindb():
 
 
 
-@transaction.commit_manually
+# @transaction.commit_manually
```
Comment (aaaaalbert) [inline common/api/maindb.py]: Fixed in e85f995e4048f10a930e653a753ee5b7e1b46752 which also fixed the module-level docstring.

File: backend/backend_daemon.py (hunk 10 above)
Comment (aaaaalbert) [inline backend/backend_daemon.py]: Addressed (but not thoroughly checked for relevance) in e85f995e4048f10a930e653a753ee5b7e1b46752.

File: backend/backend_daemon.py (hunk 11 above)
Comment (aaaaalbert) [inline backend/backend_daemon.py]: Also addressed in e85f995e4048f10a930e653a753ee5b7e1b46752, see above for caveats.

File: common/api/maindb.py [hunk 18]
```diff
@@ -575,7 +576,7 @@ def create_vessel(node, vesselname):
...
```
Comment (aaaaalbert) [inline common/api/maindb.py]: Addressed in df9e4702287f0bce53e05c437dfa53fd14c46457.

File: website/control/models.py [hunk 19]
```diff
@@ -31,36 +32,37 @@
...
-
 # First, we want to register a signal. This page recommends putting this code
 # in models.py: http://docs.djangoproject.com/en/dev/topics/signals/
 
 # Called when new database connections are created (see below).
 def _prepare_newly_created_db_connection(sender, **kwargs):
-  from clearinghouse.common.api import maindb
-  maindb.init_maindb()
+    from clearinghouse.common.api import maindb
+    maindb.init_maindb()
```
Comment (aaaaalbert) [inline website/control/models.py]: Fixed in a0b117a65e4d8531643e43fa94d037dc718908e2.

File: website/control/models.py [hunk 20]
```diff
@@ -31,36 +32,37 @@
...
-  # connection_created only exists with django >= 1.1
-  import django.db.backends.signals
-  django.db.backends.signals.connection_created.connect(_prepare_newly_created_db_connection)
+if django.VERSION >= (1.1):
+    # connection_created only exists with django >= 1.1
+    import django.db.backends.signals
+    django.db.backends.signals.connection_created.connect(_prepare_newly_created_db_connection)
 else:
-  log.error("You must use django >= 1.1 in order to support automatically " +
+    log.error("You must use django >= 1.1 in order to support automatically " +
```
Comment (aaaaalbert) [inline website/control/models.py]: Fixed in a0b117a65e4d8531643e43fa94d037dc718908e2.

File: website/control/models.py (hunk 9 above)
Comment (aaaaalbert) [inline website/control/models.py]: Fixed in a0b117a65e4d8531643e43fa94d037dc718908e2.

File: website/html/forms.py [hunk 21]
```diff
@@ -95,8 +106,9 @@ def clean_email(self):
 
 
 
-
-def gen_edit_user_form(field_list=None, *args, **kwargs):
+# Part of fix for Issue #152: change default field_list from None to '__all__'
+# for Django 1.8 compatibility
```
Comment (aaaaalbert) [inline website/html/forms.py]: 56af63031b79319663e96c5eec3cb608b7080d90 removes the comment. (I think the comment should really be the commit message for that change.)

File: website/html/views.py (hunk 6 above)
Comment (aaaaalbert) [inline website/html/views.py]: d428418330cde015ebd9d32c4c2349298bc5e783 again, comment --> commit message again.

File: website/settings.py (hunk 8 above)
Comment (aaaaalbert) [inline website/settings.py]: Addressed in 16955796fc708c38cc5b5fd5a8dc402a389405c3, but there's much more to do.

File: website/html/views.py [hunk 22]
```diff
@@ -522,7 +527,7 @@ def myvessels(request, get_form=False, action_summary="", action_detail="", remo
   
   # this user's number of donations, max vessels, total vessels and free credits
   my_donations = interface.get_donations(user)
-  my_max_vessels = interface.get_available_vessel_credits(user)	
+  my_max_vessels = interface.get_available_vessel_credits(user)
```
Comment (aaaaalbert) [inline website/html/views.py]: Addressed in 16955796fc708c38cc5b5fd5a8dc402a389405c3, but there's much more to do.

**Comment by lukpueh** [comment]
```<green light/>```

@aaaaalbert addressed the requested changes and more:
- opened issue for blanket exceptions
- removed obsolete comments
- DRYed up type override magic ** and added sensible comments
- removed legacy compatibility checking (django 1.1)
- fixed indentation
- fixed trailing whitespaces

** I think the type override magic deserves a separate ticket

Everything seems to work like before @aaaaalbert's latest changes. (I re-did the same [tests and log greping](https://github.com/SeattleTestbed/clearinghouse/pull/167#pullrequestreview-3374364) as above)

File: website/control/models.py (hunk 15 above)
Comment (aaaaalbert) [inline website/control/models.py]: Undid e9ef43b0bb44b879fdd7565d4da93d81789a2e50 in 101289d8d16d26164b4da9fe7d25cb702693cb12.

File: website/html/templates/common/geni_base.html (hunk 3 above)
Comment (aaaaalbert) [inline website/html/templates/common/geni_base.html]: Reverted in 64257db8103342b6ab3d3fedc7a7980b402b77c3.

**Pull request merged**